nixos/modules/programs/singularity.nix

```diff
@@ -84,6 +84,8 @@ in
       } // optionalAttrs cfg.enableSuid {
         enableSuid = true;
         starterSuidPath = "/run/wrappers/bin/${cfg.package.projectName}-suid";
+      } // optionalAttrs (cfg.enableNvidia && !cfg.package.supports-cdi) {
+        enableNvidiaContainerCli = true;
```

Comment:
`libnvidia-container` is not unfree so we can enable it unconditionally

Reply:
If CDI is enabled we don't want to enable wrappers, given they are not compatible.

Reply:
We don't need any wrappers, we only need to "link" singularity to `libnvidia-container` (i.e. provide with a path to nvccli)

Reply:
If there are some other packages where the choice between CDI and other backends needs to be made at build-time (rather than at runtime by the user, as I think is the case with singularity/apptainer) then we should introduce an `enable` option for CDI, not the `supports-cdi` flag. The `enable` option in that case should be disentangled from the dynamic configuration.

As far as I can see, we don't actually need any conditional logic dependent on "enabling CDI"

Reply:
All CDI specs written to the static or dynamic directories are additive. I don't see the point in an `enable` option for CDI, because even if `cdi.enable = false`, and a container runtime honors the CDI spec, it will therefore interpret the files in `/etc` and `/var/run/cdi`.

When Docker 25 is the minimum available version at nixpkgs, it's a matter of who and how a component writes to the CDI well-known directories, because all runtimes by default honor CDI: there's no point in doing `cdi.enable = false`.

I think having `cdi.static` and `cdi.dynamic` is a good balance:

- `cdi.static`: users can provide their own configuration for hardware that don't support dynamic generation; it's interesting because it rules out syntax mistakes in the JSON file.

- `cdi.dynamic`: hardware like nvidia-container-toolkit that allow the dynamic generation. Multiple providers can be supported for different (or even the same) type of hardware.

Also, mix and match of static and dynamic configurations is perfectly possible, and I think it makes everything more organized, as one could do:

```nix
cdi = {
  static = {
   provider-a = builtins.fromJSON ''
     ...
   '';
   provider-b = builtins.fromJSON ''
     ...
   '';
  };
  dynamic = {
    nvidia.enable = true;
    future-provider.enable = true;
  };
}
```

I think this is probably the best balance from a user perspective.

Reply:
> As far as I can see, we don't actually need any conditional logic dependent on "enabling CDI"

When Docker 25 is the oldest supported version of Docker, we can get rid of `supports-cdi`. There are some conditionals depending on `supports-cdi` in this PR; all of them gating `nvidia-docker`.

Reply:
What I'm trying to suggest is that we don't need `supports-cdi` and we don't need to rename `enableNvidia` - we need to _delete_ it

Reply:
> future-provider.enable = true;

Nice if we know we're going to have many of these future providers. Otherwise, `virtualisation.containers.cdi.${name}` is already long enough, and I'd avoid `virtualisation.containers.cdi.static.${name}` if I could

> `enableNvidia` ... we need to delete it

Just an impression though